proposals/p2188.md

```diff
@@ -0,0 +1,1689 @@
+following `:`.
+
+We could remove the `{field: type}` shorthand and require
+`{.field = field: type}`. This may avoid encouraging reusing the field name even
+when it is not appropriate, and would remove some syntactic sugar that's not
+formally necessary. However, we expect this to be a common case whose ergonomics
+are important.
+
+We could use a different syntax for `{field: type}` that is less divergent from
+other struct syntaxes:
+
+-   `{.field: type}` seems like a contender, but doesn't work because that is
+    already recognized as a struct type literal. Also, the use of normal binding
+    syntax means that every locally-introduced name is always introduced by
+    `name: type` where `name` is not preceded by `.`.
+-   `{.=field: type}` might be a reasonable mnemonic shorthand for
+    `{.field = field: type}`, but looks a little surprising. This is probably
+    the best choice if concerns are found with `{field: type}` syntax.
+
+### Type pattern matching
+
+We could treat type deduction as a form of pattern matching. For example, we
+could allow
+
+```
+fn F(a: Vector(T:! Type)) -> T { return a[0]; }
+fn G() -> i32 {
+  let v: Vector(i32) = (1, 2, 3);
+  // Deduces `T = i32`.
+  return F(v);
+}
+```
+
+where the value of `T` is determined by pattern-matching `Vector(T:! Type)`
+against the supplied type `Vector(i32)` of `v`. And symmetrically:
+
+```
+fn H[m: i32, n: i32](k: i32, (m, n)) { return m + n; }
+fn I() {
+  // Deduces `m = 2`, `n = 3`.
+  H(1, (2, 3));
+}
+```
+
+This would ensure consistency between pattern matching and deduction,
+potentially reducing the number of rules that Carbon developers need to learn.
+
+We find that use of pattern-matching in type position can harm readability. For
+example, the first of these two examples may be easier to read due to having
+less nesting:
+
+```
+fn F[T:! Type](x: T);
+fn F(x: (T:! Type));
+```
+
+Having distinct syntax for type-level matching and value-level matching helps
+guide the reader to the correct interpretation, even though the underlying
+matching process is expected to be similar or identical. As a result, we keep
+type deduction syntax and pattern matching syntax separate for now:
+
+-   In pattern matching, bindings and wildcards are introduced by nested `:` /
+    `:!` patterns, and the right-hand side of a binding pattern is never a
+    proper pattern.
+-   In type deduction, deduced values are specified separately and an expression
+    written in terms of those bindings describes the type.
+
+There are some contexts where there is no syntactic location for introducing
+deduced values, such as in `case` labels. For syntactic consistency, such cases
+should be addressed by adding `forall` syntax, if there is motivation to support
+deduction:
+
+```
+match (templated_value) {
+  case forall [template T:! Type] (p: T*) => { heap.Delete(p); }
+}
+```
+
+### Introducer syntax for expression patterns
```

Comment:
I feel like this section doesn't fully capture what we're giving up without this. Is it possible to capture some of the context from the discussion leading up to this Discord message? https://discord.com/channels/655572317891461132/709488742942900284/1020559500597538886

Generally, I wonder if we want to more deeply revisit whether we can and should shift the syntax in this direction. For example, I think I see plausible solutions to the purely syntactic concerns with identifying an introducer below.

That said, I'd be happy to either move that discussion to an issue for leads, and even do that as a fix-forward with even a minor note here that we anticipate revisiting this specific aspect of the syntax to see if there are approaches that might result in an overall better model.